src/tests/testClpMap.cc

```diff
@@ -0,0 +1,238 @@
+    CPPUNIT_ASSERT_EQUAL(sizeof(str), static_cast<size_t>(DefaultMemoryUsage(str)));
+    CPPUNIT_ASSERT_EQUAL(sizeof(std::string), static_cast<size_t>(DefaultMemoryUsage(std::string())));
+}
+
+void
+testClpMap::testConstructor()
+{
+    const TestMap nilA(0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilA.entries());
+
+    const TestMap nilB(0, 0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilB.entries());
+
+    const TestMap emptyC(1);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyC.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyC.entries());
+
+    const TestMap emptyD(1024);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyD.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyD.entries());
+}
+
+void
+testClpMap::testSetMemLimit()
+{
+    TestMap m(2048);
+    // overflow the map with entries to make sure it has lots of entries to purge below
+    fillMapWithElements(m, 10);
+    auto testEntriesBefore = m.entries();
+    CPPUNIT_ASSERT(testEntriesBefore > 0);
+
+    m.setMemLimit(m.memoryUsed() / 2);
+    auto entriesAfterPurge = m.entries();
+    CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
+
+    m.setMemLimit(m.memLimit() * 2);
+    // overflow the map with entries again to make sure it can grow after purging
+    fillMapWithElements(m, 10);
+    CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
+
+    // progressively shrink to size 0, checking that the number of
+    // entries decreases each time
+    testEntriesBefore = m.entries();
+    const auto singleEntrySizeEstimate = *TestMap::MemoryCountedFor("0", 0) + 10;
+    while (m.memoryUsed() > singleEntrySizeEstimate) {
+        m.setMemLimit(m.memoryUsed() / 2);
+        entriesAfterPurge = m.entries();
+        CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
+        CPPUNIT_ASSERT(entriesAfterPurge > 0);
+        testEntriesBefore = entriesAfterPurge;
+    }
+
+    // set size to 0, all entries should have been removed
+    m.setMemLimit(0);
+    CPPUNIT_ASSERT_EQUAL(size_t(0), m.entries());
+
+    // test that memory used decreases when shrinking
+    size_t curLimit = 10240;
+    size_t lastMemoryUsed;
+    m.setMemLimit(curLimit);
+    fillMapWithElements(m, 1000);
+    lastMemoryUsed = m.memoryUsed();
+    while (curLimit > 1024) // stop at any practical size
+    {
+        CPPUNIT_ASSERT(m.memoryUsed() < curLimit);
+        CPPUNIT_ASSERT(m.memoryUsed() <= lastMemoryUsed);
```

Comment:
Warning: This change request depends on other fixes in this review!

If we asked for less memory, we should get less memory (as long as there are still entries to remove).

```suggestion
        CPPUNIT_ASSERT(m.memoryUsed() < lastMemoryUsed);
```